scripts-dev/schema_versions.py

```diff
@@ -0,0 +1,143 @@
+    "synapse/storage/__init__.py",
+    "synapse/app/homeserver.py",
+)
+
+
+def get_schema_versions(tag: git.Tag) -> Tuple[Optional[int], Optional[int]]:
+    """Get the schema and schema compat versions for a tag."""
+    schema_version = None
+    schema_compat_version = None
+
+    for file in SCHEMA_VERSION_FILES:
+        try:
+            schema_file = tag.commit.tree / file
+        except KeyError:
+            continue
+
+        # We (usually) can't execute the code since it might have unknown imports.
+        if file != "synapse/storage/schema/__init__.py":
+            with io.BytesIO(schema_file.data_stream.read()) as f:
+                for line in f.readlines():
+                    if line.startswith(b"SCHEMA_VERSION"):
+                        schema_version = int(line.split()[2])
+
+                        # Bail early.
+                        break
+        else:
+            # SCHEMA_COMPAT_VERSION is sometimes across multiple lines, the easist
+            # thing to do is exec the code. Luckily it has only ever existed in
+            # a file which imports nothing else from Synapse.
+            locals: Dict[str, Any] = {}
+            exec(schema_file.data_stream.read().decode("utf-8"), {}, locals)
+            schema_version = locals["SCHEMA_VERSION"]
+            schema_compat_version = locals.get("SCHEMA_COMPAT_VERSION")
+
+    return schema_version, schema_compat_version
+
+
+def get_tags(repo: git.Repo) -> Iterator[git.Tag]:
+    """Return an iterator of tags sorted by version."""
+    tags = []
+    for tag in repo.tags:
+        # All "real" Synapse tags are of the form vX.Y.Z.
+        if not tag.name.startswith("v"):
+            continue
+
+        # There's a weird tag from the initial react UI.
+        if tag.name == "v0.1":
+            continue
+
+        try:
+            tag_version = version.parse(tag.name)
+        except version.InvalidVersion:
+            # Skip invalid versions.
+            continue
+
+        # Skip pre- and post-release versions.
+        if tag_version.is_prerelease or tag_version.is_postrelease or tag_version.local:
+            continue
+
+        tags.append((tag_version, tag))
+
+    # Sort based on the version number (not lexically).
+    return (tag for _, tag in sorted(tags, key=lambda t: t[0]))
+
+
+def calculate_version_chart() -> str:
+    repo = git.Repo(path=".")
+
+    schema_version = None
+    schema_compat_version = None
+
+    # Map of schema version -> Synapse versions at that schema.
+    schema_versions = defaultdict(list)
+
+    # Find ranges of versions which are compatible with a schema version.
+    #
+    # There are two modes of operation:
+    #
+    # 1. If schema_compat_version is None, then Synapse can only move to a new
+    #    version with schema_version >= its current version.
+    # 2. If schema_compat_version is *not* None, then Synapse can move to a new
+    #    version with schema version >= schema_compat_version.
```

Comment:
I can't say I'm grokking this at this time.

'new version' is maybe a misleading term here.

But I think №2 is saying that you can't roll back to a Synapse whose SCHEMA_VERSION is less than the `schema_compat_version` as marked in the database.

As for №1, I guess you're saying that `schema_compat_version` is assumed to be the same as `schema_version` in old versions of Synapse where there wasn't a separate concept of `schema_version` — I find that a little bit surprising: does that mean there were boundaries where you couldn't roll back even one version? (I also seem to remember being here when `SCHEMA_COMPAT_VERSION` was added and I thought from [very fallible] memory that it made things *more* restrictive than before, not *less*). Don't suppose you have a reference from old-synapse handy?

(But also: I am willing to just trust you on this)

Reply:
> I find that a little bit surprising: does that mean there were boundaries where you couldn't roll back even one version?

Yes, exactly. See #9933. I think your explanations are fairly reasonable though, we have:

№1 (pre-schema_compat_version) where you can only rollback to the same schema_version as you were on (but can roll forward to any newer schema_version).
№2 (post-schema_compat_version) where you can rollback to anything >= schema_compat_version (but can roll forward to any newer schema_version).

Reply:
Hopefully a714262827d8fbe40b34675dac15ff4456f1d4d4 clarified this.